staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,232 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watch
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog"
+)
+
+// WatcherFunc is a function that is responsible for creating a watcher starting at sinceResourceVersion.
+type WatcherFunc func(sinceResourceVersion string) (watch.Interface, error)
+
+// resourceVersionGetter is an interface used to get resource version from events.
+// We can't reuse an interface from meta otherwise it would be a cyclic dependency and we need just this one method
+type resourceVersionGetter interface {
+	GetResourceVersion() string
+}
+
+// RetryWatcher will make sure that in case the underlying watcher is closed (e.g. due to API timeout or etcd timeout)
+// it will get restarted from the last point without the consumer even knowing about it.
+// RetryWatcher does that by inspecting events and keeping track of resourceVersion.
+// Especially useful when using watch.Until where premature termination is causing issues and flakes.
+// Please note that this is not resilient to etcd cache not having the resource version anymore - you would need to
+// use Informers for that.
+type RetryWatcher struct {
+	lastResourceVersion string
+	watcherFunc         WatcherFunc
+	resultChan          chan watch.Event
+	stopChan            chan struct{}
+	doneChan            chan struct{}
+}
+
+// NewRetryWatcher creates a new RetryWatcher.
+// It will make sure that watcher gets restarted in case of recoverable errors.
+// The initialResourceVersion will be given to watchFunc when first called.
+func NewRetryWatcher(initialResourceVersion string, watcherFunc WatcherFunc) *RetryWatcher {
+	rw := &RetryWatcher{
+		lastResourceVersion: initialResourceVersion,
+		watcherFunc:         watcherFunc,
+		stopChan:            make(chan struct{}),
+		doneChan:            make(chan struct{}),
+		resultChan:          make(chan watch.Event, 0),
+	}
+
+	go rw.receive()
+	return rw
+}
+
+func (rw *RetryWatcher) send(event watch.Event) bool {
+	// Writing to an unbuffered channel is blocking operation
+	// and we need to check if stop wasn't requested while doing so.
+	select {
+	case rw.resultChan <- event:
+		return true
+	case <-rw.stopChan:
+		return false
+	}
+}
+
+// doReceive returns true when it is done, false otherwise.
+// If it is not done the second return value holds the time to wait before calling it again.
+func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
+	watcher, err := rw.watcherFunc(rw.lastResourceVersion)
+	if err != nil {
+		klog.Errorf("retryWatcher: watcherFunc failed: %v", err)
+		// Retry
+		return false, 0
```

Comment:
won't this hot-loop retry on network errors?

Reply:
It will, but it will always check for the stop channel so the timeout (context) from Until cancels it. This used to hard fail, but in this last iteration I have decided for approach more similar to informers. I am  open to other views on how we should treat errors here, but my latest thinking is proceed at all cost, unless you know for sure you can't - network errors, permissions it all can be transilient

Reply:
added artificial waits for hard failures